### ISSUE TITLE
How do I run UpdateTool GUI: 1.0.0?

### ISSUE THREAD

**Issue by viking2** [mynttt/UpdateTool#49]
I installed OpenJDK11U-jdk_x64_windows_hotspot_11.0.10_9.msi and downloaded  updatetool-gui-win-1.0.0.jar.
How to I run updatetool-gui-win-1.0.0.jar? Double-clicking the file only opens it as an archive (in my Filemanager).
What program do I "open"/run it with?

EDIT: 
I tried to run it from the command prompt like this, but that didn't work:
C:\.....\jdk-11.0.10.9-hotspot\bin\java.exe E:\.....\updatetool-gui-win-1.0.0.jar
Error: Could not find or load main class E:\.....\updatetool-gui-win-1.0.0.jar
Caused by: java.lang.ClassNotFoundException: E:\....\updatetool-gui-win-1.0.0.jar

**Comment by mynttt** [comment]
Seems like the  *.jar extension is associated with your archive program. To change this use Right Click -> Properties and set the file extension to be associated with Java.

Else you can run it by cmd via `java -jar nameofjar.jar`.

**Comment by viking2** [comment]
A little progress..
I ran this:
C:\Windows\system32>java -jar D:\xxx\updatetool-gui-win-1.0.0.jar
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[INFO ] - 2021-01-25 01:03:08 @ Config.load: Loading config: C:\Users\xxx\AppData\Roaming\UpdateToolGUI

Even with the error, the GUI opened: 
Set proper Java 11+ binary (java / java.exe).
Set correct plex media server folder (the one with 'Plug-in Support' in it).
First time downloading UpdateTool.jar
Downloaded version: 1.4.7

It is not clear what the plex media server folder should be. 
Should it be this path: D:\P_Data\Plex Media Server\Plug-in Support\
or just this D:\P_Data\Plex Media Server\?

It is also not clear what the Java 11+ binary (java / java.exe) path should be .
I tried C:\xxx\jdk-11.0.10.9-hotspot\bin 
and also C:\xxx\jdk-11.0.10.9-hotspot\
Neither seemed to work. I got the message "Set proper Java 11+ binary (java / java.exe)."

Where was the "Downloaded version: 1.4.7" downloaded? What do I do with it? Am I supposed to run this instead of the original updatetool-gui-win-1.0.0.jar?

Assuming that I set the paths correctly, and I use the correct xxx.jar version, what would I do after that? 
Do I close it? Will it still run every N hours?
Do I need to keep it open to run every N hours? 
Can I run it manually to update the IMDb ratings?

**Comment by mynttt** [comment]
Don't worry about the warning in the log file it's not going to cause any trouble! The folder should be the "Plex Media Server" one and the Java executable should be the jdk...\bin\java.exe. The downloaded version is stored in a special data folder that can be reached by clicking the top left button, it will update automatically. You don't have to do anything with it as the tool will mange the versions for you. 

Basically once everything is set correctly you can just click start and the GUI will run the tool and output the log in the text window. You can then either leave it running forever or stop it by clicking Stop and closing the GUI again. The tool would be invoked every n hours as long as it's running. 

You might also need to fill out the API key sections of your libraries are TV Shows or use the TMDB agent.

**Comment by viking2** [comment]
1. My main purpose of your Tool is to get the IMDb rating for TV-shows. The Plex agents do not support this.
It is not clear to me if how to setup the Agents for TV-Show and Movies in Plex so that your Tool will work? Doesn't it work independent of the Agents?

2. I have tried another Add-on for Kodi "Light IMDb Ratings Update" https://forum.kodi.tv/showthread.php?tid=316342. That one works great, and doesn't need any API keys as it is scraping from IMDb instead. I guess that you haven't considered that method...?

3. As you stated, the TVDB now needs a subscription for their API key (and I don't have a legacy key), so I will skip that.

4.  I didn't understand what you meant by "..fill out the API key sections of your libraries are TV Shows or use the TMDB agent." ? Do I need a TMDB API key to get the IMDb rating for TV-shows (and Movies)?

5. To get the TMDB API key, I need to fill out a form where I need to specify the Application Name, URL and Summary!? What do I enter for this? Do I enter "Update Tool", https://github.com/mynttt/UpdateTool, and a summary? They also ask for my name and address.......(Why??)

**Comment by mynttt** [comment]
1. The agents have different ways of storing the identity of the matched item and thus need to be treated differently as it is necessary to resolve their identity and match it to the IMDB item before querying the IMDB dataset for a rating.

2. Kodi most likely stores the IMDB id's directly which the TVDB/TMDB Plex agents don't do. You can't just take the TVDB id and query IMDB with it as it will yield no results.

4.  Should be "...fill out the API key sections if your libraries are TV Shows or use the TMDB agent." had a typo in there.

5. Get the developer one and just fill out something, they won't contact you as all their keys are automatically approved.

**Comment by viking2** [comment]
Thanks again for your answers. Unfortunately, the IMDb ratings were not downloaded:
Error = "No libraries found. ":
(I have checked "Ignore All Movie Libraries" but not "Ignore TV Show Libraries")
```
Started UpdateTool:

[INFO ] - 2021-01-26 18:27:42 @ Main.loadImplementation: Loaded implementation << imdb-docker << updatetool.imdb.ImdbDockerImplementation
[INFO ] - 2021-01-26 18:27:42 @ Main.main: Running version: 1.4.7
[INFO ] - 2021-01-26 18:27:42 @ Main.main: Args: {imdb-docker={schedule=12}}
[INFO ] - 2021-01-26 18:27:42 @ Main.main: << INIT: imdb-docker @ {schedule=12} >>
[INFO ] - 2021-01-26 18:27:42 @ Main.testApiTmdb: Testing TMDB API key: 53f14048e79c3880ed272b64ab1459f5
[INFO ] - 2021-01-26 18:28:21 @ Main.genericApiTest: Test passed. API Key is valid.
[INFO ] - 2021-01-26 18:28:21 @ ImdbDockerImplementation.bootstrap: TMDB API key enabled TMDB <=> IMDB matching. Will process TMDB backed Movie and TV Series libraries and TMDB orphans.
[INFO ] - 2021-01-26 18:28:21 @ ImdbDockerImplementation.bootstrap: No TVDB API authorization string detected. Will process TVDB backed TV Series libraries.
[INFO ] - 2021-01-26 18:28:21 @ ImdbDockerImplementation.bootstrap: Starting IMDB Watchdog
[INFO ] - 2021-01-26 18:28:21 @ ImdbDockerImplementation.bootstrap: Plex data dir: D:\@P_Data\Plex Media Server
[INFO ] - 2021-01-26 18:28:21 @ ImdbDockerImplementation.bootstrap: Capabilities: [TMDB, NO_MOVIE, VERBOSE_XML_ERROR_LOG]
[INFO ] - 2021-01-26 18:28:21 @ Main.main: << INIT SUCCESS >>
[INFO ] - 2021-01-26 18:28:21 @ Scheduler.go: Scheduler is loading tasks... Blocking until completely set-up and ready to go.
[INFO ] - 2021-01-26 18:28:21 @ Scheduler.go: Scheduling tasks...
[INFO ] - 2021-01-26 18:28:21 @ Scheduler.lambda$go$0: Scheduled imdb-docker task to run @ every 12 hour(s).
[INFO ] - 2021-01-26 18:28:21 @ Scheduler.lambda$go$1: Queued task imdb-docker for immediate execution.
[INFO ] - 2021-01-26 18:28:21 @ Scheduler.go: Running supplied tasks immediately NOW!
[INFO ] - 2021-01-26 18:28:21 @ TaskWrapper.run: ================================================
[INFO ] - 2021-01-26 18:28:21 @ TaskWrapper.run: Starting task: imdb-docker | Execution count: 0
[INFO ] - 2021-01-26 18:28:21 @ TaskWrapper.run: ================================================
[INFO ] - 2021-01-26 18:28:22 @ ImdbDockerImplementation$ImdbBatchJob.run: Library IDs on ignore list: []
[INFO ] - 2021-01-26 18:28:22 @ ImdbDockerImplementation$ImdbBatchJob.run: No libraries found. Sleeping until next invocation...
[INFO ] - 2021-01-26 18:28:22 @ TaskWrapper.run: ================================================
[INFO ] - 2021-01-26 18:28:22 @ TaskWrapper.run: Suggesting JVM to run the GC as soon as possible (Request might be ignored!)
[INFO ] - 2021-01-26 18:28:22 @ TaskWrapper.run: Completed imdb-docker in 0.437025s. - Invoking next task or going to sleep. It is safe to suspend execution if no other task is being invoked immediately.
```
Here is how I setup my Show Agents:
![](https://i.gyazo.com/97affc718cc5947b3fb97a0c4b9c1f42.png)
and, but probably not relevant since I don't have the TVDB API Key?:
![](https://i.gyazo.com/3ea026f8e8abfbc1611d4a0499de62ed.png)

p.s. After I signed up for the API key, I got a message:
_"We recently launched our TV API and would love some feedback if you have a chance to check it out."_
Were you aware? Can this be useful?

**Comment by viking2** [comment]
I haven't heard back. Any idea why the Update tool doesn't work for TV-shows (at least not when using the TMDB API key)? Could you please try to reproduce?

Thank you (Danke vielmals)!

**Comment by mynttt** [comment]
I think the issue is that all of your libraries are matched via the TVDB agent. If you later change it to TMDB it won't update the actual match of each existing item it will only apply TMDB metadata to newly added items.

So the reason why no library is found is because they all use TVDB and no TVDB API authorization is given. The TMDB TV API is already supported but that doesn't help in your case as you would require TVDB access.

**Comment by viking2** [comment]
I never understood how the Agents worked or what the purpose it would be to have TVDB for the TMDB Agent and vice versa. My setup was more or less random (no idea really how to set it up properly)... I also don't know what type info is downloaded from each Agent...

In any case, is there is no solution? Is there no way for me now to get IMDb rating using the TMDB Api Key? What about changing the Agents (How?) and do a Metadata refresh? 

If that wouldn't work, can I create a new database? Maybe if I call it TV-shows2 and point to the same directories? How would I setup the Agents properly to get it to work?

**Comment by mynttt** [comment]
Basically in the Plex Database there is an unique identifier for each item, for TVDB/TMDB items it looks approximately like this:

TMDB: tmbd://1234/1/12
TVDB: tvdb://528/2/12

This associates the entry with either the TMDB via ID 1234 for Season 1 and Episode 12 or via TVDB via ID 528 for Season 2 and Episode 12. UpdateTool now takes this information and contacts either TVDB/TMDB to get the IMDB ID in the format tt12345678. This ID is then used to lookup the current rating in an from IMDB acquired dataset.

To do this resolvement the matching API needs to be connected i.e. you can't resolve a TVDB item via the TMDB API and vice versa. In your case it is likely that all TV Show items have the TVDB identifier and as far as I know even a refresh metadata will not change it! You could try to create a second library and set the Agent in your screenshot to The Movie Database and untick the "TheTVDB" entry. I don't know how good the quality of TMDB is for TV Shows tho and as far as I know it has only sometimes been used as a fallback in my own library.

I hope that this helps you a bit if you have more questions just let me know!

**Comment by viking2** [comment]
I created an empty folder and set it up as a TV-show library using TMDB. I then moved in 3 TV shows; The Blacklist, Batwoman & Resident Alien. Only The Black list got an IMDb rating but not the other shows. If use the "Recommended Tab"(whatever that means) and click on the The Black List, I see the ratings:
![](https://i.gyazo.com/b183fdc7def1a4c45bf7ead1a0f55602.png)

However, if use the "Library Tab" there is no IMDb rating!?:
![](https://i.gyazo.com/5727f31fe838b9ae5c24b18a6b441e41.png)

Do you know if there is some way that I could get a temporary TVDB API Key so that I can test how that would work?

Here is the log file:
```
Started UpdateTool:

[INFO ] - 2021-02-01 19:36:38 @ Main.loadImplementation: Loaded implementation << imdb-docker << updatetool.imdb.ImdbDockerImplementation
[INFO ] - 2021-02-01 19:36:38 @ Main.main: Running version: 1.4.7
[INFO ] - 2021-02-01 19:36:38 @ Main.main: Args: {imdb-docker={schedule=12}}
[INFO ] - 2021-02-01 19:36:38 @ Main.main: << INIT: imdb-docker @ {schedule=12} >>
[INFO ] - 2021-02-01 19:36:38 @ Main.testApiTmdb: Testing TMDB API key: xxxxxxxxxxxx
[INFO ] - 2021-02-01 19:36:40 @ Main.genericApiTest: Test passed. API Key is valid.
[INFO ] - 2021-02-01 19:36:40 @ ImdbDockerImplementation.bootstrap: TMDB API key enabled TMDB <=> IMDB matching. Will process TMDB backed Movie and TV Series libraries and TMDB orphans.
[INFO ] - 2021-02-01 19:36:40 @ ImdbDockerImplementation.bootstrap: No TVDB API authorization string detected. Will process TVDB backed TV Series libraries.
[INFO ] - 2021-02-01 19:36:40 @ ImdbDockerImplementation.bootstrap: Starting IMDB Watchdog
[INFO ] - 2021-02-01 19:36:40 @ ImdbDockerImplementation.bootstrap: Plex data dir: D:\@P_Data\Plex Media Server
[INFO ] - 2021-02-01 19:36:40 @ ImdbDockerImplementation.bootstrap: Capabilities: [TMDB, NO_MOVIE, VERBOSE_XML_ERROR_LOG]
[INFO ] - 2021-02-01 19:36:40 @ Main.main: << INIT SUCCESS >>
[INFO ] - 2021-02-01 19:36:40 @ Scheduler.go: Scheduler is loading tasks... Blocking until completely set-up and ready to go.
[INFO ] - 2021-02-01 19:36:40 @ Scheduler.go: Scheduling tasks...
[INFO ] - 2021-02-01 19:36:40 @ Scheduler.lambda$go$0: Scheduled imdb-docker task to run @ every 12 hour(s).
[INFO ] - 2021-02-01 19:36:40 @ Scheduler.lambda$go$1: Queued task imdb-docker for immediate execution.
[INFO ] - 2021-02-01 19:36:40 @ Scheduler.go: Running supplied tasks immediately NOW!
[INFO ] - 2021-02-01 19:36:40 @ TaskWrapper.run: ================================================
[INFO ] - 2021-02-01 19:36:40 @ TaskWrapper.run: Starting task: imdb-docker | Execution count: 0
[INFO ] - 2021-02-01 19:36:40 @ TaskWrapper.run: ================================================
[INFO ] - 2021-02-01 19:36:41 @ ImdbDockerImplementation$ImdbBatchJob.run: Library IDs on ignore list: []
[INFO ] - 2021-02-01 19:36:41 @ ImdbRatingDatasetFactory.requestSet: IMDB Dataset has the timestamp: 1611774942217 and violates the update every 86400000 ms constraint. Refreshing dataset...
[INFO ] - 2021-02-01 19:36:41 @ ImdbRatingDatasetFactory.downloadData: Downloading IMDB rating set from: https://datasets.imdbws.com/title.ratings.tsv.gz
[INFO ] - 2021-02-01 19:36:43 @ ImdbRatingDatasetFactory.downloadData: Download succeeded @ ./__tmp_rating.gz
[INFO ] - 2021-02-01 19:36:43 @ ImdbRatingDatasetFactory.extractData: Extracting dataset...
[INFO ] - 2021-02-01 19:36:44 @ ImdbRatingDatasetFactory.extractData: Extraction completed.
[INFO ] - 2021-02-01 19:36:44 @ ImdbRatingDatasetFactory.readData: Reading data...
[INFO ] - 2021-02-01 19:36:48 @ ImdbRatingDatasetFactory.readData: 1118767 lines read.
[INFO ] - 2021-02-01 19:36:48 @ ImdbDockerImplementation$ImdbBatchJob.run: [SERIES] TEST TV Shows IMDb has 4 item(s)
[INFO ] - 2021-02-01 19:36:48 @ ImdbDockerImplementation$ImdbBatchJob.run: Processing [SERIES] TEST TV Shows IMDb with UUID 0322ba32-dca9-4809-92ce-70522ec52fb4 at stage: CREATED
[INFO ] - 2021-02-01 19:36:48 @ ImdbPipeline.analyseDatabase: Resolving IMDB identifiers for items. Only warnings and errors will show up...
[INFO ] - 2021-02-01 19:36:48 @ ImdbPipeline.analyseDatabase: Items that show up here will not be processed by further stages of the pipeline.
[WARN ] - 2021-02-01 19:36:48 @ DefaultResolvement.resolve: Item: 'Prior Criminal History' has no matching IMDB resolver and will be ignored. (guid=com.plexapp.agents.thetvdb://361315/2/2?lang=en)
[WARN ] - 2021-02-01 19:36:48 @ DefaultResolvement.resolve: Item: 'Bat Girl Magic!' has no matching IMDB resolver and will be ignored. (guid=com.plexapp.agents.thetvdb://361315/2/3?lang=en)
[WARN ] - 2021-02-01 19:36:48 @ DefaultResolvement.resolve: Item: 'Pilot' has no matching IMDB resolver and will be ignored. (guid=com.plexapp.agents.thetvdb://368166/1/1?lang=en)
[INFO ] - 2021-02-01 19:36:48 @ ImdbPipeline.analyseDatabase: Filtered 4 invalid item(s).
[INFO ] - 2021-02-01 19:36:48 @ ImdbPipeline.transformMetadata: Transforming 0 item(s)
[INFO ] - 2021-02-01 19:36:48 @ ImdbPipeline.transformMetadata: Transformed entries for 0 items(s).
[INFO ] - 2021-02-01 19:36:48 @ ImdbPipeline.updateDatabase: Nothing to update. Skipping...
[INFO ] - 2021-02-01 19:36:48 @ ImdbPipeline.updateXML: Updating XML fallback files for 0 item(s).
[INFO ] - 2021-02-01 19:36:49 @ ImdbPipeline.updateXML: Completed updating of XML fallback files.
[INFO ] - 2021-02-01 19:36:49 @ ImdbDockerImplementation$ImdbBatchJob.run: Job returned PASS : Job finished correctly
[INFO ] - 2021-02-01 19:36:49 @ ImdbDockerImplementation$ImdbBatchJob.run: Job finished successfully for [SERIES] TEST TV Shows IMDb with UUID 0322ba32-dca9-4809-92ce-70522ec52fb4
[INFO ] - 2021-02-01 19:36:49 @ ImdbDockerImplementation$ImdbBatchJob.run: Completed batch successfully.
[INFO ] - 2021-02-01 19:36:49 @ TaskWrapper.run: ================================================
[INFO ] - 2021-02-01 19:36:49 @ TaskWrapper.run: Suggesting JVM to run the GC as soon as possible (Request might be ignored!)
[INFO ] - 2021-02-01 19:36:49 @ TaskWrapper.run: Completed imdb-docker in 8.2134698s. - Invoking next task or going to sleep. It is safe to suspend execution if no other task is being invoked immediately.
```

**Comment by mynttt** [comment]
Hi,

could be that with the TMDB agent only an IMDB match could be resolved with that specific episode but not with the root element of the series. That's why only the episode has the rating associated. You could look into the file `cache-tmdbseriesBlacklist.json` that probably has the ID for the root element of the series. It would retry again after a few days in case TMDB has updated the entry.

You could get a TVDB key by paying for a month and use a legacy key (V3) as V4 is not supported right now (they're still transitioning from V3 to V4 and not all features have been implemented yet...........)

**Comment by viking2** [comment]
But none of the other series has any ratings at all....?

I am guessing that maybe the only way to currently get IMDb rating for TV shows in Plex is to use a v3 TVDB key and your tool?
  
However, I don't want to pay for the TVDB key unless that will work with the update tool (I have no other use for it).
Also, I believe that the v3 key will stop working in a month or so, and it is unknown if v4 key will work. 

The alternative is to go over to Kodi, with the Light IMDb Ratings Update Addon (https://forum.kodi.tv/showthread.php?tid=316342). I can then get TV show ratings without any key.

**Comment by mynttt** [comment]
Well I have no idea why that might be the case. My guess is no IMDB association for the items in the TMDB database.

You're likely correct, I'm using a v3 key (got it when they still were free) and everything is working well. I don't know what will happen with the v4 API. As of right now it is still incomplete and does not offer the same features as the V3 API. They did not do a good transition and their backlog is still having tons of uncompleted items (https://support.thetvdb.com/kb/faq.php?id=74).

You can transition to Kodi if that solves this issue but there is no other way for you to get these ratings in Plex. That's just how things work over at Plex.
